### PR TITLE
PoC: Investigate support for database

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR explores [SQModel](https://github.com/tiangolo/sqlmodel) integration within the Lightning App framework to simplify database manipulation.

# Why SQModel

From [Sebastián Ramírez](https://github.com/tiangolo), SQLModel removes boilerplate to interact with database from Python. Its SQLModel is a SQLAlchemy model and a Pydantic model ✨ Which makes it very easy to use with FastAPI. 

My personal opinion is that it removes most of the boilerplate to manipulate the specs from control-plane to data plane and reconciliation.

carbon - 2022-08-05T125538 466

https://user-images.githubusercontent.com/12861981/183063574-63e517ed-009b-4adc-b920-9dc3d815ac35.mov

https://user-images.githubusercontent.com/12861981/183063620-8bc5cc6b-7483-4ba1-a0e4-819e48613c84.mov

Another benefit is the ease to access reading and writing from / to the database for the users.

## Reading

```python
class FlowDB(LightningFlow):

    def run(self):
        if self.schedule("...."):
                 self.read_from_db()

   def read_from_db(self):
        with Session(self.engine) as session:
              ....
              return ...
```

## Writing to the DB through the Rest API

Associated with the `configure_api` PR https://github.com/Lightning-AI/lightning/pull/13945, you get full pydantic validation + database entries.

```python
class FlowDB(LightningFlow):

    def run(self):
        if self.schedule("...."):
                 self.read_from_db()

   def something_handler(self, config: SomethingConfig):
        with Session(self.engine) as session:
              ....
              return ...

   def configure_api(self):
         return [Post("/create_something", self.something_handler)
```

Fixes #\<issue_number>

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/tests_app/core/test_lightning_flow.py [hunk 1]
```diff
@@ -634,3 +637,28 @@ def run(self):
...
+        if self.counter.counter >= 1000:
+            self._exit()
+
+        print(self.counter.counter)
+        self.counter.counter += 1
+
+
+def test_lightning_flow_sql_model():
+
+    app = LightningApp(FlowDB(), database="sqlite")
```
Comment (hhsecond) [inline tests/tests_app/core/test_lightning_flow.py:663]: passing the `database` argument to `LightningApp` is very disconnected from initializing the DB object inside the Flow. I don't have a solution to recommend for that but we might need to reconsider

Comment (manskx) [reply tests/tests_app/core/test_lightning_flow.py:663]: I think we can avoid passing it to LightningApp. the `self.db` property could be a subclass of a baseDB and contain the needed info for the db.

We can do something like this ?

```python


class LightningBaseDB:
     .....

class LightningDefaultSqliteDB(LightningBaseDB):
    # stored as a file in the drive
    .....

class LightningPostgresComponent(LightningBaseDB):
   .....


class Table1(Present):
    id: Optional[int] = Field(default=None, primary_key=True)
    count: int = 0

class Table2(Present): 
    id: Optional[int] = Field(default=None, primary_key=True)
    name: str

class MySimpleDB(LightningBaseDB):
    # schema
    table1: Table1
    table2: Table2


class MyApp(LightningFlow):
    def __init__(self):
        super().__init__()

        self.db = MySimpleDB()
        self.db.reload()

    def run(self):
        if self.db.tabel1.counter >= 1000:
            self._exit()

        print(self.counter.counter)
        self.db.tabel1.counter += 1


def test_lightning_flow_sql_model():

    app = LightningApp(FlowDB())

```

Comment (tchaton) [reply tests/tests_app/core/test_lightning_flow.py:663]: Hey @hhsecond @manskx.

I will think about the proposition above. IMO, it isn't exactly what users need.

They need a simple way to persist a component state to a database in an atomic way and be able to restart from that state. In this PoC PR, the state is persisted with the component name and can be reloaded.

I like being able to access the tables in a structured way, but the counter is raw within the table, not the table entirely.

Comment (tchaton) [reply tests/tests_app/core/test_lightning_flow.py:663]: @lantiga @awaelchli What are your thoughts? I would be curious to see someone else PoC on this subject. I think it is one of the most important API to ace from the framework.

Comment (lantiga) [reply tests/tests_app/core/test_lightning_flow.py:663]: I think we still don’t have clarity on the purpose.

I see two disjoint needs here, we need to tackle them as such, although at some point we will see the two needs addressed by the same solution, but it’s too early to say:
- providing a way for users to persist information (not state), for instance jobs they historically launched. When I say “not state” I mean that state might still hold a reference to a particular snapshot in time of such db, as a whole.
- providing a durable recording of state and its history, potentially with transactional semantics, a-la temporal.io

I strongly recommend we take this PR as a design brainstorm. We are not ready to commit to a design or the other.

I’ll make more specific comments asap.

File: src/lightning_app/db/spec.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    SQLModel = object
+
+
+class LightningSpec(SQLModel):
+    __table_args__ = {"extend_existing": True}
+    _engine: Optional[Any] = None
+    component_name: Optional[str]
+    _should_reload: bool = False
+
+    def __setattr__(self, key, value) -> None:
```
[the remark this answers is not in the document]
Comment (tchaton) [reply src/lightning_app/db/spec.py:17]: Yes, exactly. I am actually thinking an even cleaner design.

File: src/lightning_app/db/spec.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+if _is_sql_model_available():
+    from sqlmodel import select, Session, SQLModel
+else:
+    SQLModel = object
+
+
+class LightningSpec(SQLModel):
+    __table_args__ = {"extend_existing": True}
+    _engine: Optional[Any] = None
```
[the remark this answers is not in the document]
Comment (tchaton) [reply src/lightning_app/db/spec.py:13]: Yes, but this is primarily meant as a utility mechanism. However, I have a another design idea which won't create this linking.

**Comment by justusschock** [comment]
@tchaton is this still relevant or can this be closed? Research Studio does have databases, right?

**Pull request closed without merging**